### ISSUE TITLE
Support of jarjar'd Guava and GSON libraries

### ISSUE THREAD

**Issue by before** [immutables/immutables#171]
We are using jarjar'd libraries of Guava and GSON, since we have third party libraries that use older versions of them. Our versions have an additional package to encapsulate Guava and GSON properly (e.g. `custom.com.google.common.collect`).

Is it possible to configure *immutables* to prefer these libraries instead of the original ones in our classpath? Would be great to set this as an annotation processor option (see [SupportedOptions](http://docs.oracle.com/javase/7/docs/api/javax/annotation/processing/SupportedOptions.html)).

Is there a workaround for now that I could use, e.g. using [JarJar](https://code.google.com/p/jarjar/) to migrate *value-2.0.21*?

**Comment by elucash** [comment]
You would better have third party libraries repackaged with older version and use plain newer version in your code, in that way you can repackage your version also while building jar.

Currently there's no support for that and I also cannot see how you can change it in adhoc way aside from using plain `com.google.common.collect` in your code and repackage it afterwards.
I'll see how we can parametrize this, i.e. guava package name in generate sources.
Also you can force generation of JDK-only code by using `Style(jdkOnly=true)`, but this is questionable alternative.
Thank you for the issue report!

**Comment by before** [comment]
Sadly we use for a longer time these repackaged libraries and I have no chance to change that at the moment. It would great to have an option for it. I'm sure there are other projects that use shaded versions of Guava and GSON as well. Thank you for your time

**Comment by trask** [comment]
Try using the maven-shade-plugin to shade guava in your
Immutables-generated classes, and let us know if this works for you.

Thanks,
Trask

**Comment by elucash** [comment]
Another way it could be implemented is via hook-point to apply arbitrary source replacements on the generated sources before they are written to files. We're currently post processing files to add imports from fully qualified names, a step could be added to make replacemens specified somewhere, probably in the text file on a classpath. Quite interesting if there are any other compelling use-cases for the functionality if implemented?

**Comment by before** [comment]
@trask Sadly we have a legacy Ant based build and it is not easy to use *maven-shade-plugin* there without extensive hacks.

@elucash A hook-point would also be an interesting solution which provides us more flexibility. Could it be implemented as a function that can added to `Value.Style`?

```Java
@Value.Style(importsPostProcessing = new Function<Import, Import>() {
    public Import apply(Import input) {
        // change import if necessary
    }
})
```

**Comment by elucash** [comment]
@before it doesn't work exactly this way in java annotations ). But it's a good idea! And it could be implemented as extension on the classpath visible during annotation processing, the one that have a separate class with no-arg constructor and is referenced in `META-INF/services/<provider.classname>` files. It does mean that you'll have to compile it to a separate jar and put on the same classpath as annotation processor: it might be separate for some build tools or special build configurations.
I'll sketch something for the nearest release.

**Comment by before** [comment]
@elucash thank you. I'm looking forward to it. BTW, it is a great utility and we really enjoy it.

**Comment by elucash** [comment]
The solution is ad-hoc service provider `org.immutables.extgenerator.GeneratedImportsModifier`

```java
package org.immutables.extgenerator;

/**
 * Ad-hoc tool to change imports in generated source files after imports have been processed.
 * The annotation processor with load implementations of this
 * interface using classpath service loading, see {@link ServiceLoader} for the details.
 * Make sure it will be present of the annotation processor classpath rather than compile classpath
 * if they are not the same in your build tool. Implementations should take special care of safety,
 * i.e. they should neither throw exception during construction or processing nor leave set of
 * imports in an improperly modified state.
 * @since 2.1
 */
public interface GeneratedImportsModifier {
  /**
   * Modify imports set after import processing/extraction but before writing generated files.
   * Important to note that there are no way to change classnames as other parts of source file
   * refers to class by simple name. However package-name part of imports may be rewritten.
   * @param packageOfGeneratedFile informative package name of the generated file, may be used to
   *          employ different strategies for different packages.
   * @param imports modifiables set of imports
   */
  void modify(String packageOfGeneratedFile, Set<String> imports);
}
```

Service provider implementation together with `/META-INF/services...` entry should be then available on the annotation processing classpath: either regular compile classpath or on a special annotation processing classpath if it's running in isolated one.

See sample test implementation [.../PreconditionsRewriter.java](https://github.com/immutables/immutables/blob/master/value-fixture/src/org/immutables/fixture/generatorext/PreconditionsRewriter.java)

**Comment by elucash** [comment]
Released in 2.1.0

**Comment by before** [comment]
@elucash thank you for this. Which file do you expect under `/META-INF/services...` based on your sample test? How would it look like in detail?

**Comment by elucash** [comment]
It's convention used to load extensions from classpath. In particular `java.util.ServiceLoader` use it and details about the convention are available there in javadoc.
In nutshell, you need to create file in classpath (i.e. inside jar) named `META-INF/services/org.immutables.extgenerator.GeneratedImportsModifier` for the interface name, which should contain a single line (or more lines if you will have many implementations, which is unlikely). This line should specify fully qualified name of extension provider class, as in my previous example it would be `org.immutables.fixture.generatorext.PreconditionsRewriter`, substitute it with your class. This class should be public with public no-arg constructor.

* `META-INF/services/org.immutables.extgenerator.GeneratedImportsModifier`
```
org.immutables.fixture.generatorext.PreconditionsRewriter
```
hope it helps

**Comment by before** [comment]
@elucash Put everything together and it works like a dream. Thank you very much.

**Comment by before** [comment]
@elucash After some testing I miss one thing. The generator does not understand that we use Guava types, because they lay in another package.
```Java
import custom.com.google.common.base.Optional;
import custom.com.google.common.collect.ImmutableList;
import org.immutables.value.Value;
// Define abstract value type
@Value.Immutable
public interface ValueObject {
  String name();
  ImmutableList<Integer> counts();
  Optional<String> description();
}
```
For example, the generated `ValueObject.Builder` does not work with `ImmutableList.Builder<Integer>` for counts and the description is not absent by default. Should I create a new ticket for that feature or is there something to compensate this?

**Comment by elucash** [comment]
Indeed, at some point I also thought about this, but it somehow escaped from the focus. I'm reopening issue as existing solution is not of much use, at least for the case of repackaged Guava.

**Comment by before** [comment]
@elucash I worked around this limitation by exchanging the `UnshadeGuava` implementation in the packaged JAR. Additionally I added a generator extension which manipulated the imports to use our package prefix as well. Not really comfortable, but as a workaround fine.

**Comment by elucash** [comment]
Next release will add ability to specify guava package prefix (both for recognition and code generation) as a system property "guava.prefix" during compilation. So I hope it should be slightly easier than or at least alternative to your smart workaround.

Here's maven example
```xml
<plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.3</version>
        <compilerArgs>
            ...
            <arg>-Dguava.prefix=com.my.shaded.google.common</arg>
          </compilerArgs>
      </plugin>
```

**Comment by elucash** [comment]
released in 2.1.1

**Comment by before** [comment]
@elucash thank you for this improvement. How does it work for the GSON library? Is there also a plan to have a prefix? At the moment it looks like I need to JarJar the package for the import `com.google.gson`.

**Comment by elucash** [comment]
could do same for gson
`-Dguava.prefix=...`
will it help?

**Comment by before** [comment]
Yes, that would fit well.

**Comment by before** [comment]
@elucash I have issues to modify the _Ant_-based build to pass the argument through the annotation processor. It results in the error `javac: invalid flag: -Dguava.prefix=myprefix.com.google.common`). I tried it also with the prefix `-J` (regarding to the documentation of [javac -Joption](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javac.html#BHCFFJBJ)), but than I get `javac: invalid flag: -J-Dguava.prefix=myprefix.com.google.common`.

It seems that an annotation processor should be fed in a another way with options to ensure proper handling (see [accessing VM arguments during compile time](http://stackoverflow.com/questions/14520851/accessing-vm-arguments-during-compile-time)).

Have you thought or tried to use the [SupportedOptions](http://docs.oracle.com/javase/7/docs/api/javax/annotation/processing/SupportedOptions.html)?

**Comment by before** [comment]
@elucash sorry, that I bothered you. I resolved it by setting a property task and forking the javac task and it works now as expected.
```XML
<property name="guava.prefix" value="myprefix.com.google.com" />
<javac fork="true" ...
```

**Comment by elucash** [comment]
no problem, sorry for the delayed response. At the time I had to choose how to configure it, it was obvious that while annotation processor options was kinda "appropriate" way to do this, but having to change lifecycle of couple classes to pass along the configuration, so system properties seemed like easier and arguably not so bad solution.

I glad you find a way to configure it for Ant!
Added `gson.prefix` system property in just released v2.1.7.
